release.py

```diff
@@ -117,20 +123,34 @@ def update_docker_images(version_tag, update_latest_tag):
     run_or_die(clean_cmd)
 
     # OSS app
-    push_image('//deployment:release_onprem', image_tag=version_tag)
+    oss_tag = version_tag
+    enterprise_tag = 'enterprise-'+version_tag
+    pushed_oss_image = False
+    pushed_enterprise_image = False
+    pushed_enterprise_exec_image = False
+    if not gcr_tag_exists("flame-public/buildbuddy-app-onprem", oss_tag):
+        push_image('//deployment:release_onprem', image_tag=oss_tag)
+        pushed_oss_image = True
     # Enterprise app
-    push_image('//enterprise/deployment:release_enterprise', image_tag='enterprise-'+version_tag)
+    if not gcr_tag_exists("flame-public/buildbuddy-app-enterprise", enterprise_tag):
+        push_image('//enterprise/deployment:release_enterprise', image_tag=enterprise_tag)
+        pushed_enterprise_image = True
     # Enterprise executor
-    push_image('//enterprise/deployment:release_executor_enterprise', image_tag='enterprise-'+version_tag)
+    if not gcr_tag_exists("flame-public/buildbuddy-executor-enterprise", enterprise_tag):
+        push_image('//enterprise/deployment:release_executor_enterprise', image_tag=enterprise_tag)
+        pushed_enterprise_exec_image = True
 
     # update "latest" tags
     if update_latest_tag:
         # OSS app
-        push_image('//deployment:release_onprem', image_tag='latest')
+        if pushed_oss_image:
```

Comment:
if this script fails on e.g.  line 136, then when we retry this script, `pushed_oss_image` will be false since the `gcr_tag_exists` check will return false, and we won't update the latest tag on line 147, but should.

one thing that could work is to use the docker API for each of these 3 images to resolve the image digests for both the vX.Y.Z tag and the 'latest' tag, and if the digests are mismatched, then pull the vX.Y.Z tag and re-tag+push as 'latest' (pull+push would be much faster than re-building, and guarantees that the rebuild doesn't result in a new digest if the bazel build isn't 100% reproducible from the preceding failed run, which should be rare but can happen)

Reply:
Looks like we can also do it without pushing/pulling: https://stackoverflow.com/a/72908550

Reply:
Fixed